clang/runtime/dpct-rt/include/dpct/bindless_images.hpp

```diff
@@ -117,6 +137,341 @@ class image_mem_wrapper {
+      <sycl::ext::oneapi::experimental::resource_win32_handle> ext_mem_desc{
+        ext_mem_win_handle,
+        sycl::ext::oneapi::experimental::external_mem_handle_type::
+            win32_nt_dx12_resource,
+        _res_size_bytes};
+
+    _interop_mem_handle_ptr =
+      new sycl::ext::oneapi::experimental::interop_mem_handle;
+    *_interop_mem_handle_ptr =
+      sycl::ext::oneapi::experimental::import_external_memory(
+        ext_mem_desc, get_default_queue());
+
+    uniqueD3D11Res.insert(_d3d11_res);
+  }
+
+  ~interop_mem_wrapper() {
+    assert(_interop_mem_handle_ptr && "Resource is not registered!");
+    
+    CloseHandle(_win_nt_handle);
+
+    sycl::ext::oneapi::experimental::release_external_memory(
+        get_interop_mem_handle(), get_default_queue());
+    
+    uniqueD3D11Res.erase(_d3d11_res);
+  }
+#endif
+
+  interop_mem_wrapper(const interop_mem_wrapper &) = delete;
+
+  interop_mem_wrapper &operator=(const interop_mem_wrapper &) = delete;
+
+  const sycl::ext::oneapi::experimental::interop_mem_handle
+  get_interop_mem_handle() const noexcept {
+    return *_interop_mem_handle_ptr;
+  }
+
+  sycl::ext::oneapi::experimental::image_mem_handle
+  get_img_mem_handle() const noexcept {
+    return *_img_mem_handle_ptr;
+  }
+
+  void *
+  get_buf_ptr() const noexcept {
+    return _buf_ptr;
+  }
+
+  sycl::ext::oneapi::experimental::image_descriptor
+  get_img_desc() const noexcept {
+    return _img_desc;
+  }
+
+  size_t get_res_size_bytes() {
+    return _res_size_bytes;
+  }
+
+  bool is_buffer_res() const {
+    return _res_is_buffer;
+  }
+
+  void map_resource(sycl::queue q) {
+    if (_res_is_buffer) {
+      assert(!_buf_ptr && "Resource is already mapped!");
+
+      // NOTE: 'map_external_linear_memory' API is not yet merged by Codeplay
+      /*
+      _buf_ptr =
+        sycl::ext::oneapi::experimental::map_external_linear_memory(
+          get_interop_mem_handle(), _img_desc, q);
+      */
+      throw std::runtime_error("Mapping a buffer resoure is not yet supported!");
+    }
+    else {
+      assert(!_img_mem_handle_ptr && "Resource is already mapped!");
+
+      _img_mem_handle_ptr =
+        new sycl::ext::oneapi::experimental::image_mem_handle;
+      *_img_mem_handle_ptr =
+        sycl::ext::oneapi::experimental::map_external_image_memory(
+          get_interop_mem_handle(), _img_desc, q);
+      
+      _img_mem_wrapper_ptr = new image_mem_wrapper(this);
+    }
+  }
+
+  image_mem_wrapper* get_mapped_mipmapped_array() {
+    assert(!_res_is_buffer &&
+        "Buffer resouce cannot be accessed as an array!");
+    assert(_img_mem_wrapper_ptr && "Resouce is not mapped!");
+
+    return _img_mem_wrapper_ptr;
+  }
+
+  image_mem_wrapper* get_sub_resource_mapped_array(
+      unsigned array_index, unsigned mip_level) {
+    assert(!_res_is_buffer &&
+        "Buffer resouce cannot be accessed as an array!");
+    assert(array_index == 0 &&
```

Comment:
Same as above.

Reply:
This check is added to honor CUDA's condition [here](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__INTEROP.html#group__CUDART__INTEROP_1g0dd6b5f024dfdcff5c28a08ef9958031:~:text=resource%20is%20mapped.-,If%20resource%20is%20not%20a%20texture%20then%20it%20cannot%20be%20accessed%20via%20an%20array%20and%20cudaErrorUnknown%20is%20returned,-.%20If%20arrayIndex%20is)